### ISSUE TITLE
Shortcomings of Characters Table

### ISSUE THREAD

**Issue by PullFrom** [w3c/alreq#49]
Hi,

in section [Punctuations and symbols](http://w3c.github.io/alreq/#punctuations-and-symbols) I find that these three symbols are marked as _not_ being used in Arabic but only in Farsi:

U+066A  ARABIC PERCENT SIGN
U+066B  ARABIC DECIMAL SEPARATOR
U+066C  ARABIC THOUSANDS SEPARATOR

Which characters are used for that purpose in Arabic documents? Just the Western/European percent, comma, and point?

Any advice very much appreciated,
- Michael

**Comment by khaledhosny** [comment]
- U+066A is definitely used in Arabic with Arabic-Indic numbers.
- U+066B should be used as well, but it is not commonly use. Most likely because it is not available in common keyboard layout, so people usually use comma or even a small U+0631(I have seen that mainly in news papers, probably it used because it looks more closer to U+066B than a comma is).
- U+066C was used in old documents (I have books circa 1930 that use it) haven’t seen it in modern use, but I don’t even remember seeing thousands separated at all.

**Comment by khaledhosny** [comment]
Looking in that section, I see more oddities as well.
- U+00AB and U+00BB are the preferred quotation marks in Arabic, but usually a rounded variant is used (that looks like small (( and )) rather than << and >>).
- U+00D7 and U+00F7, U+0640, U+2026 and U+2212 are used in Arabic of course.

**Comment by khaledhosny** [comment]
I think the whole characters section needs revision, what is the best way to do that?

**Comment by r12a** [comment]
@khaledhosny create a new issue, detailing your suggestions, or send a pull request (see http://w3c.github.io/i18n-activity/editorial-guidelines/ for advice on how to do that, if you need it).

**Comment by r12a** [comment]
@khaledhosny it may also help things along if you are able to attend a telecon, and discuss the issues there. (There's one this afternoon.)

**Comment by khaledhosny** [comment]
OK.
Will see if I can attend the telecon as well.

**Comment by PullFrom** [comment]
Thanks a lot for the feedback, which is – luckily – along what I heard from the customer’s translator person: No thousands separator, but Arabic-Indic digits and as a consequence I added U+066A percent sign and U+066B decimal separator. Since we are preprocessing numbers automatically we can easily use the correct decimal separator.

**Comment by r12a** [comment]
> U+00AB and U+00BB are the preferred quotation marks in Arabic, but usually a rounded variant is used (that looks like small (( and )) rather than << and >>).

here are some examples.  The middle title seems to have a similar, single version: 
![arabic-guillemet-cropped](https://cloud.githubusercontent.com/assets/4839211/15674769/7b0e8f6a-2736-11e6-8fc8-23171c90f868.jpg)

**Comment by mostafah** [comment]
@mmh-docufy-de @khaledhosny Thank you for the feedback.

We discussed this in today’s teleconference. We built these tables by combining the exemplar characters from ICU4J and CLDR for each language. These sources may not always be complete or up-to-date. For Persian, a national Iranian standard exists and we added characters listed in that too. For Arabic, unfortunately, we didn’t found a similar source.

We have to stick with resources and can’t add new characters to the list merely based on expert knowledge, but whenever there is a controversy about a character we can add a note for that one in the tables to say about that.

What do you think about this solution?

Of course, whenever a new reliable resource becomes available we would be happy to add that to improve the tables.

Again, thanks for the feedback. As I said, we can use these discussions to mark the potential flaws in the tables. So, please, keep them coming. :-)

**Comment by khaledhosny** [comment]
I think ICU and CLDR are not reliable sources for character usage if they report such widely used characters not used in Arabic. The Arabic percent sign for example is the only percent sign used in Egypt, the same about all math operators reported not used, Arabic keyboard layouts even has × and ÷.

Arabic is used in many countries compared to Persian and there is no single official source, not to mention regional difference which can vary greatly. If we insist on only official sources, so we need much softer language than “used” and “not used” because the current tables are outright wrong in many cases.

**Comment by mostafah** [comment]
How are those keyboard layouts defined? Can they be used as a source?

**Comment by khaledhosny** [comment]
I have no idea how they are defined, but most likely they were inherited from typewriters and early computer systems in a way or another, just like the QWERTY layout. There is no standard body that defines Arabic keyboard layouts. I think they make a good start as a source, but they do not cover everything since typesetters are not limited by keyboard layouts.

**Comment by mostafah** [comment]
If they help minimize the exceptions, I agree with using them. But let’s see what others think about it.

**Comment by PullFrom** [comment]
How about the Microsoft Windows keyboard layouts, the may be considered a de facto standard? I found this: https://msdn.microsoft.com/en-us/goglobal/bb964651

**Comment by khaledhosny** [comment]
Microsoft layout does not show Arabic-Indic numbers and associated symbols as Windows converts ASCII digits to localised ones on the fly based on the locale (at least in Arabic), so they keyboard layout has 1…9%, etc, but based on locale what you will be seeing in Windows is ١…‎٩٪.

**Comment by r12a** [comment]
@mostafah i guess i came out of the meeting with a slightly different interpretation of what we planned to do. My understanding was that we should add to the table characters we believe are needed for arabic support, but make it clear where the evidence for that comes from.  So i think it's ok to add ٪, for example, but we'd need a note or indication somewhere to say that it's not indicated as needed by CLDR, but that we see a need for it.

**Comment by mostafah** [comment]
@ntounsi has also said this in PR #50:

> ARABIC LETTER TCHEH, PEH, JEH and others are marked o. Should be x. Not used in Arabic. 
> Arabic Maddah, Hamza above, Hamza bellow ... are marked x. Should be o or dark bullet.
> So are punctuation and symbols
> « » ÷ x ARE used in Arabic.

Copying here because it’s related to this discussion.

**Comment by mostafah** [comment]
@r12a Oh, seems like I misunderstood our consensus. Apologies for that.

And maybe I misinterpreted that because this was my own preference! I prefer the rigor of not adding characters to the tables unless they are in a documented source, but note the flaws of the tables in the complementary text.

But I can go with the other way if the consensus is on that.

And as a small recap, I guess these are the two actions we have here:
- [ ] See if reliable keyboard layouts can be found for Arabic and update the tables with them
- [ ] Add the missing characters reported in this issue either to the tables or only as notes

**Comment by brawer** [comment]
> I think ICU and CLDR are not reliable sources for character usage

If something is broken, can you file a bug report here? http://unicode.org/cldr/trac/newticket

> keyboard layouts

By the way, CLDR collects data with keyboard layouts for various operating systems
([description](http://cldr.unicode.org/index/charts/keyboards), [data files](http://www.unicode.org/repos/cldr/trunk/keyboards/)). Again, if something is wrong, please don’t hesitate to file [bug reports](http://unicode.org/cldr/trac/newticket) so that Unicode’s data can improve.

**Comment by ntounsi** [comment]
In Morocco, the comma is used as decimal separator.  
Thousands separator is space or dot. The picture shows a document  with  a same column containing amounts  and a sum. Amounts with space separator, and the sum with dot separator.
![untitled5](https://cloud.githubusercontent.com/assets/15004708/15904230/ada9b444-2d9f-11e6-9b66-52524c9838c8.png)
![untitled6](https://cloud.githubusercontent.com/assets/15004708/15904231/adb5e7d2-2d9f-11e6-9bb3-d7656b7cca5a.png)
 In the same document, we see other amount 1000 without thousands separator.
![untitled2](https://cloud.githubusercontent.com/assets/15004708/15904266/cf6ef9a4-2d9f-11e6-8789-6aed9163fcdc.png)
BTW, the % sign is used also here. At the left of the number, but this is not a problem both % 25 or 25 % are used indifferently.
 Another story in the next picture
![untitled3](https://cloud.githubusercontent.com/assets/15004708/15904280/dfdad72c-2d9f-11e6-82f4-554c8b84440f.png)
 Use of 1° 2° as item list style. Note also the double angle quotation mark

Regards,
Najib

**Comment by brawer** [comment]
Regarding CLDR being unreliable, I’ve filed a ticket at http://unicode.org/cldr/trac/ticket/9492. Please tell us (Unicode) what exact changes should be done to the exemplar characters data.

**Comment by r12a** [comment]
Najib mentions that % may appear before or after the number it qualifies. Is this a widespread occurrence?  And does it also apply for ٪ ?

**Comment by r12a** [comment]
Since this issue has grown into 'what's wrong with the character table', let me note here that all the control characters except the last one (FEFF) are needed for Arabic too.

**Comment by PullFrom** [comment]
For me, born and raised in a LTR world, the terms "before" and "after" are synonyme to "left" and "right" – until one learns XSL-FO where they have a clear direction-dependent meaning.

If it is not totally clear from the context, I would suggest to make it clear if "before" means "RTL before" (= appearing to the right of other glyphs) or "LTR before". 

For e.g. "٪٢٥" I would say, that in Arabic text the percent sign comes after the number.

**Comment by khaledhosny** [comment]
With Arabic-Indic digits, the percent (٪) is always to the right of the number, with BiDi rules this means it comes after the number in the text stream.

**Comment by ntounsi** [comment]
A rapid look, I found the (%) sign sometime after sometime before the number. 
In an official printed Moroccan document. It is the normal (%) signed which is used. See pictures.
![bo1](https://cloud.githubusercontent.com/assets/15004708/15976131/6f4440aa-2f40-11e6-89d6-204df6e0394f.png)
![bo2](https://cloud.githubusercontent.com/assets/15004708/15976132/6f5080fe-2f40-11e6-9fd3-750a2e635ae6.png)

About the Arabic percent sign (٪), it is "less used". However, I found some other document with both signs. See pictures.
![im2](https://cloud.githubusercontent.com/assets/15004708/15976148/7bab9334-2f40-11e6-807b-332b752b5ce4.png)
![im1](https://cloud.githubusercontent.com/assets/15004708/15976149/7bafbc5c-2f40-11e6-894d-cf0cb015bdb1.png)

**Comment by mostafah** [comment]
We talked about the place of percent sign in a teleconference and were curious about what the school books in Iran do. I checked it out and apparently they put it on the right side of the number.

![screen shot 1395-04-22 at 14 12 06](https://cloud.githubusercontent.com/assets/531910/16762756/28719fec-483b-11e6-8111-87f11e418e08.png)

**Comment by ntounsi** [comment]
Some authors may put a space between the number and the sign, and its place can then depend on the base direction.  Especially this kind of numbers + sign are likely to appear alone in a table cell.

**Comment by shervinafshar** [comment]
I would like to summarize the discussion regarding shortcomings of character tables to make them actionable (either on ALReq side or CLDR's). For those of you who are inclined regarding the positioning of percent sign, please do open another issue and discuss the matter there. Thank you.

Here are the summary of the issues and actions needed to be taken:

1. Add double angle-quotes, U+00AB and U+00BB (« and ») to CLDR Arabic exemplar set for punctuation;
_1.1. Question:_ is that if there is any common usage of the single angle-quotes, U+2039 and U+203A (‹ and ›) for nested quotations or alike?
2. Re: U+066A (ARABIC PERCENT SIGN), U+066B (ARABIC DECIMAL SEPARATOR), 
U+066C (ARABIC THOUSANDS SEPARATOR), ALReq tables need to include the data provided in CLDR [Number Symbols](http://unicode.org/reports/tr35/tr35-numbers.html#Number_Symbols) to resolve the discrepancies for Arabic. No action needed on the data for CLDR;
3. Re: U+00D7 (MULTIPLICATION SIGN), ALReq tables need to include the CLDR Number Symbols data (see item 2);
4. Re: U+00F7 (DIVISION SIGN), CLDR doesn’t have a dedicated data item (according to UTS#35) to record that for each locale. ALReq included this in the charts for Persian based on ISIRI keyboard layout standards. Need to be discussed in CLDR TC to see if a data item to record such symbol might be possible. Otherwise, discuss in ALReq on how to proceed including it;
5. Re: U+0640 (ARABIC TATWEEL), CLDR data doesn’t include it at the moment for either Arabic or Persian. ALReq tables included it in charts for Persian based on ISIRI standards (see item 4). Propose to CLDR to add U+0640 to auxiliary exemplar sets for both Persian and Arabic;
6. Re: U+2026 (HORIZONTAL ELLIPSIS), CLDR data includes it in exemplar punctuation for Persian, but not for Arabic. Propose to CLDR to add U+2026 to Arabic punctuation exemplar as it’s already in use for ellipsis patterns;
7. Re: U+2212 (MINUS SIGN), Arabic data from CLDR under Number Symbols shows that Arabic is using U+002D (HYPHEN-MINUS) both for `latn` and `arab` number systems.    ALReq need to include this code-point in the tables;
_7.1. Question:_ Is CLDR data in this regard acceptable and denotes common usage?

@mostafah, I think we need to revise the CLDR lookup in the script generating the charts to go down in the numbers > symbols sections of the data (or get it from ICU4J).

@khaledhosny, @ntounsi, @mmh-docufy-de; please review and comment on my summary. After we gathered the comments, there would be an update on the ticket that @brawer kindly filed on our behalf.

Thanks to all involved in this discussion.

**Comment by behnam** [comment]
About keyboard layouts, CLDR already has a collection of vendor-defined layouts: http://cldr.unicode.org/index/charts/keyboards

Some of the layouts are based on standards and are reliable, like `fa-t-k0-osx-isiri.xml`.

Some of the layouts have problems or are not up-to-date, like `fa-t-k0-android.xml`.

So, look like, at least for now, we have to manually cross-check that source with our tables.

**Comment by mostafah** [comment]
@shervinafshar Thanks for the effort of going through the issues and cleaning up the discussion.

I updated the tables in the PR #89 by adding number symbols from ICU4J. This adds these characters:

- U+00D7 MULTIPLICATION SIGN
- U+066A ARABIC PERCENT SIGN
- U+066B ARABIC DECIMAL SEPARATOR
- U+066C ARABIC THOUSANDS SEPARATOR

for Arabic. That solves point 2 and 3 from Shervin’s comment above.

It also adds U+221E INFINITY for both languages.

But U+002D HYPHEN-MINUS is not added for Arabic because we are excluding ASCII from our tables and have stated this in the document. IIRC we made that decision to avoid polluting our tables with ASCII punctuations that are common among Arabic and Latin. These are the ASCII characters that are being dropped because of this rule: ` !"'()*+,-./:;<=>?[\]{|}` (and space). I’m thinking that we might want to reconsider this.
